serverless/cmd/clone2serverless/main.go

```diff
@@ -0,0 +1,209 @@
+	vkey       = flag.String("log_vkey", "", "Verifier key for the log checkpoint signature.")
+	origin     = flag.String("log_origin", "", "Expected Origin for the log checkpoints.")
+)
+
+func main() {
+	flag.Parse()
+
+	if len(*mysqlURI) == 0 {
+		glog.Exit("Missing required parameter 'mysql_uri'")
+	}
+	if len(*origin) == 0 {
+		glog.Exit("Missing required parameter 'log_origin'")
+	}
+	if len(*vkey) == 0 {
+		glog.Exit("Missing required parameter 'log_vkey'")
+	}
+	if len(*outputRoot) == 0 {
+		glog.Exit("Missing required parameter 'output_root'")
+	}
+
+	db, err := logdb.NewDatabase(*mysqlURI)
+	if err != nil {
+		glog.Exitf("Failed to connect to database: %v", err)
+	}
+	logVerifier, err := note.NewVerifier(*vkey)
+	if err != nil {
+		glog.Exitf("Failed to instantiate Verifier: %q", err)
+	}
+
+	ctx := context.Background()
+	toSize, inputCheckpointRaw, _, err := db.GetLatestCheckpoint(ctx)
+	if err != nil {
+		glog.Exitf("Failed to get latest checkpoint from clone DB: %v", err)
+	}
+	logRoot, fromSize, err := initStorage(logVerifier, *outputRoot)
+	if err != nil {
+		glog.Exitf("Could not initialize tile-based storage: %v", err)
+	}
+
+	// Set up a background thread to report progress (as it can take a long time).
+	rCtx, rCtxCancel := context.WithCancel(ctx)
+	defer rCtxCancel()
+	r := reporter{
+		lastReported: fromSize,
+		lastReport:   time.Now(),
+		treeSize:     toSize,
+	}
+	go func() {
+		ticker := time.NewTicker(5 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				r.report()
+			case <-rCtx.Done():
+				return
+			}
+		}
+	}()
+
+	results := make(chan logdb.StreamResult, 16)
+	glog.V(1).Infof("Sequencing leaves [%d, %d)", fromSize, toSize)
+	go db.StreamLeaves(ctx, fromSize, toSize, results)
+
+	index := fromSize
+	for result := range results {
+		if result.Err != nil {
+			glog.Exitf("Failed fetching leaves from DB: %v", result.Err)
+		}
+		workDone := r.trackWork(index)
+		err := logRoot.Assign(ctx, index, result.Leaf)
+		if err != nil {
+			glog.Exitf("Failed to sequence leaves: %v", err)
+		}
+		workDone()
+		index++
+	}
+	rCtxCancel()
+
+	glog.V(1).Infof("Integrating leaves [%d, %d)", fromSize, toSize)
+	if _, err := log.Integrate(ctx, fromSize, logRoot, rfc6962.DefaultHasher); err != nil {
+		glog.Exitf("Failed to integrate leaves: %v", err)
+	}
+	glog.V(1).Info("Writing checkpoint")
+	if err := logRoot.WriteCheckpoint(ctx, inputCheckpointRaw); err != nil {
+		glog.Exitf("Failed to write checkpoint: %v", err)
+	}
+	glog.V(1).Infof("Successfully created tlog with %d leaves at %q", toSize, *outputRoot)
+}
+
+// initStorage loads or creates the tile-based log storage in the given output directory
+// and returns the storage, along with the tree size currently persisted.
+func initStorage(logVerifier note.Verifier, outputRoot string) (*fs.Storage, uint64, error) {
+	var logRoot *fs.Storage
+	var fromSize uint64
+	if fstat, err := os.Stat(outputRoot); os.IsNotExist(err) {
```

Comment:
I think this misses the case where `err != nil` but does not represent a non-existent file - it _might_ not matter since we'll likely just error out lower down when we try to open the log files, but maybe explicitly logging & bailing might be better?

Reply:
Good call, added this check in.